src/clients/c/tb_client.h

```diff
@@ -249,68 +263,85 @@ typedef enum TB_PACKET_STATUS {
     TB_PACKET_INVALID_DATA_SIZE = 7,
 } TB_PACKET_STATUS;
 
-typedef struct tb_packet_t {
-    void* user_data;
-    void* data;
-    uint32_t data_size;
-    uint16_t tag;
-    uint8_t operation;
-    uint8_t status;
-    uint8_t reserved[32];
-} tb_packet_t;
+typedef enum TB_INIT_STATUS {
+    TB_INIT_SUCCESS = 0,
+    TB_INIT_UNEXPECTED = 1,
+    TB_INIT_OUT_OF_MEMORY = 2,
+    TB_INIT_ADDRESS_INVALID = 3,
+    TB_INIT_ADDRESS_LIMIT_EXCEEDED = 4,
+    TB_INIT_SYSTEM_RESOURCES = 5,
+    TB_INIT_NETWORK_SUBSYSTEM = 6,
+} TB_INIT_STATUS;
+
+typedef enum TB_CLIENT_STATUS {
+    TB_CLIENT_OK = 0,
+    TB_CLIENT_INVALID = 1,
+} TB_CLIENT_STATUS;
 
-typedef void* tb_client_t; 
+typedef enum TB_REGISTER_LOG_CALLBACK_STATUS {
+    TB_REGISTER_LOG_CALLBACK_SUCCESS = 0,
+    TB_REGISTER_LOG_CALLBACK_ALREADY_REGISTERED = 1,
+    TB_REGISTER_LOG_CALLBACK_NOT_REGISTERED = 2,
+} TB_REGISTER_LOG_CALLBACK_STATUS;
 
-typedef enum TB_STATUS {
-    TB_STATUS_SUCCESS = 0,
-    TB_STATUS_UNEXPECTED = 1,
-    TB_STATUS_OUT_OF_MEMORY = 2,
-    TB_STATUS_ADDRESS_INVALID = 3,
-    TB_STATUS_ADDRESS_LIMIT_EXCEEDED = 4,
-    TB_STATUS_SYSTEM_RESOURCES = 5,
-    TB_STATUS_NETWORK_SUBSYSTEM = 6,
-} TB_STATUS;
+typedef enum TB_LOG_LEVEL {
+    TB_LOG_ERR = 0,
+    TB_LOG_WARN = 1,
+    TB_LOG_INFO = 2,
+    TB_LOG_DEBUG = 3,
+} TB_LOG_LEVEL;
 
 // Initialize a new TigerBeetle client which connects to the addresses provided and
 // completes submitted packets by invoking the callback with the given context.
-TB_STATUS tb_client_init(
-    tb_client_t* out_client,
+TB_INIT_STATUS tb_client_init(
+    tb_client_t *client_out,
     const uint8_t cluster_id[16],
-    const char* address_ptr,
+    const char *address_ptr,
     uint32_t address_len,
-    uintptr_t on_completion_ctx,
-    void (*on_completion)(uintptr_t, tb_client_t, tb_packet_t*, uint64_t, const uint8_t*, uint32_t)
+    uintptr_t completion_ctx,
+    void (*completion_callback)(uintptr_t, tb_packet_t*, uint64_t, const uint8_t*, uint32_t)
 );
```

Comment:
To get this to 90%, let's also pass in an allocator? This won't allow us to get rid of `libc` _immediately_, due to `@cImport`, but will keep API compatible with that. 

I suggest using Lua's version of allocator:


```
typedef void * (*lua_Alloc) (void *ud,
                             void *ptr,
                             size_t osize,
                             size_t nsize);
```

The `ud` would be `completion_ctx`, `osize` is old size, `nsize` is new size. 

https://www.lua.org/manual/5.3/manual.html

Reply:
Decided not do do this: we need mmap to init io-uring anyway, so might as well use it for allocation.